File: csharp/nucleotide-count/NucleotideCount.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

public static class NucleotideCount
{
    public static IDictionary<char, int> Count(string sequence)
    {
        if (InvalidSequence(sequence))
            throw new ArgumentException(
                $"'{sequence}' is an invalid sequence.", "sequence");

        return $"{sequence}ACGT"
            .GroupBy(
                nucleotide => nucleotide
            )
            .ToDictionary(
                group => group.Key,
                group => group.Count() - 1
            );
    }
    private static bool InvalidSequence(string sequence)
    {
        var pattern = "[^ACGT]";
        return Regex.IsMatch(sequence, pattern);
    }
}

File: fsharp/nucleotide-count/NucleotideCount.fs
module NucleotideCount
open System.Text.RegularExpressions

let nucleotideCounts (strand: string): Option<Map<char, int>> =
    let badStrand = Regex(@"[^ACGT]")
    let countNucl (acc: Map<char, int>) char = acc.Add(char, acc[char]+1)
    let empty = Map([('A', 0); ('C', 0); ('G', 0); ('T', 0)])
    match strand with
    | _ when badStrand.IsMatch(strand) -> None
    | _ -> (empty, strand) ||> Seq.fold countNucl |> Some